internal/handler/handler.go

```diff
@@ -0,0 +1,349 @@
+package handler
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"regexp"
+	"strconv"
+
+	"github.com/VTerenya/employees/internal"
+	"github.com/VTerenya/employees/internal/service"
+	"github.com/shopspring/decimal"
+)
+
+var (
+	positionsRe      = regexp.MustCompile(`[\\/]positions\?limit=(\d+)&offset=(\d+)$`)
+	positionRe       = regexp.MustCompile(`[\\/]position[\\/](\S+)$`)
+	createPositionRe = regexp.MustCompile(`[\\/]position$`)
+
+	employeesRe      = regexp.MustCompile(`[\\/]employees\?limit=(\d+)&offset=(\d+)$`)
+	employeeRe       = regexp.MustCompile(`[\\/]employee[\\/](\S+)$`)
+	createEmployeeRe = regexp.MustCompile(`[\\/]employee$`)
+)
+
+type Handler struct {
+	service service.Serv
+}
+
+func NewHandler(service service.Serv) *Handler {
+	return &Handler{service: service}
+}
+
+func (h *Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("content-type", "application/json")
+	switch {
+	case r.Method == http.MethodGet && positionsRe.MatchString(r.URL.String()):
+		h.getPositions(w, r)
+		return
+	case r.Method == http.MethodGet && employeesRe.MatchString(r.URL.String()):
+		h.getEmployees(w, r)
+		return
+
+	case r.Method == http.MethodGet && positionRe.MatchString(r.URL.String()):
+		h.getPosition(w, r)
+		return
+	case r.Method == http.MethodGet && employeeRe.MatchString(r.URL.String()):
+		h.getEmployee(w, r)
+		return
+
+	case r.Method == http.MethodPost && createPositionRe.MatchString(r.URL.Path):
+		h.createPosition(w, r)
+		return
+	case r.Method == http.MethodPost && createEmployeeRe.MatchString(r.URL.Path):
+		h.createEmployee(w, r)
+		return
+
+	case r.Method == http.MethodDelete && positionRe.MatchString(r.URL.String()):
+		h.deletePosition(w, r)
+		return
+	case r.Method == http.MethodDelete && employeeRe.MatchString(r.URL.String()):
+		h.deleteEmployee(w, r)
+		return
+
+	case r.Method == http.MethodPut && createPositionRe.MatchString(r.URL.String()):
+		h.updatePosition(w, r)
+		return
+	case r.Method == http.MethodPut && createEmployeeRe.MatchString(r.URL.String()):
+		h.updateEmployee(w, r)
+		return
+	default:
+		fmt.Println("error: not found")
+		http.NotFound(w, r)
+		return
+	}
+}
+
+func (h *Handler) getPositions(w http.ResponseWriter, r *http.Request) {
+	query, err := url.ParseQuery(r.URL.RawQuery)
+	if err != nil {
+		http.Error(w, "error with server", http.StatusInternalServerError)
+		return
+	}
+	limit, err := strconv.ParseInt(query["limit"][0], 10, 64)
+	if err != nil {
+		http.Error(w, "error with server", http.StatusInternalServerError)
+		return
+	}
+	offset, err := strconv.ParseInt(query["offset"][0], 10, 64)
+	if err != nil {
+		http.Error(w, "error with server", http.StatusInternalServerError)
+		return
+	}
+	positions, err := h.service.GetPositions(int(limit), int(offset))
+	if err != nil {
+		if err.Error() == "incorrect data" {
```

Comment:
Better to define list of errors you want to compare with in the separate `errors` package and then use `errors.Is()` instead of `err.Error()` to compare an errors

Reply:
+